synapse/http/server.py

```diff
@@ -492,6 +492,38 @@ class RootOptionsRedirectResource(OptionsResource, RootRedirect):
     pass
 
 
+class _JsonProducer:
+    """
+    Iteratively write JSON to the request.
+    """
+
+    def __init__(self, request, json_encoder, json_object):
+        self.request = request
+        self._generator = json_encoder.iterencode(json_object)
+
+    def start(self):
+        self.request.registerProducer(self, False)
+
+    def resumeProducing(self):
+        # We've stopped producing in the meantime.
+        if not self.request:
+            return
+
+        # Get the next chunk and write it to the request. Calling write will
+        # spin the reactor (and might be re-entrant).
```

Comment:
Is this true? We're' not awaiting here so the reactor should not be able to spin?

Reply:
I think it's cut-and-pasted from elsewhere; but I seem to remember that it *is* true that this can be re-entrant.

Reply:
This was originally cut-and-pasted from the [`NoRangeStaticProducer`](https://github.com/twisted/twisted/blob/twisted-20.3.0/src/twisted/web/static.py#L768-L779).

Reply:
I added some logging and I actually wasn't able to cause this to be re-entrant?

Anyway the question of "how does the reactor spin?" is convoluted, but I think I found the answer:
* Calling `registerProducer` on the `Request` actually [registers it with the `HTTPChannel`](https://github.com/twisted/twisted/blob/1b439fbabc584f56c7bf6abf754395ba1ed3800b/src/twisted/web/http.py#L985-L987)
* The `HTTPChannel` then [converts the `IPullProducer` to an `IPushProducer`](https://github.com/twisted/twisted/blob/1b439fbabc584f56c7bf6abf754395ba1ed3800b/src/twisted/web/http.py#L2547-L2554).
* It also calls [`startStreaming` which generates a `cooperator`](https://github.com/twisted/twisted/blob/414d490bb7811fb7b55cf2cfbf71f7a8b8527485/src/twisted/internet/_producer_helpers.py#L85-L91)
* The cooperated task (`_PullToPush._pull`) [yields after each call to `resumeProducing`](https://github.com/twisted/twisted/blob/414d490bb7811fb7b55cf2cfbf71f7a8b8527485/src/twisted/internet/_producer_helpers.py#L65-L82) to give up control to the reactor.

The result is that in one reactor "tick" this should generate <~1 KB of data.

Reply:
Ok, right interesting. Since this function is not a coroutine I don't think this can yield back to the reactor half way through the function

Reply:
Right, it "yields" when the function runs off, NOT in the middle of the function.

Do you think any changes are needed here?

Reply:
I guess I still don't really see what this comment is trying to convey. `write` may or may not cause the reactor to wake up immediately, and I don't see why we are trying to highlight that.

If we want to highlight that this function may be re-entrant, it'd be good to mention what the code is doing to make that safe (I guess something to do with the fact that the write is the last thing we do in the function, and its only during `write` that we can become re-entrant?)

Reply:
I believe it is important that you check if `self._request` still exists since something could have called `stopProducing`.

Maybe it isn't useful to clarify -- I can remove the part of the comment if you'd like (or maybe move it above the `self._request` check?)

Reply:
Saying that at by the check I think would be best

Reply:
@erikjohnston I clarified a bunch of comments! Let me know if it is better now!